### DISCUSSION TITLE
model.setBlock() method error msg: Not supported for TensorFlow Engine

### DISCUSSION THREAD

**Discussion by iamuspace** [deepjavalibrary/djl#2410] [Q&A] [answered]
Q: When I run my DJL simple code, error occurs. 

OS : Windows 10 
CPU : AMD 64bit 
GPU : Nvidia Titan X 
Engine : TensorFlow 

Error when model.setBlock() method.

std output of code : 

2023-02-19 16:13:44.781483: I external/org_tensorflow/tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-02-19 16:13:45.206304: I external/org_tensorflow/tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 10773 MB memory:  -> device: 0, name: NVIDIA GeForce GTX TITAN X, pci bus id: 0000:21:00.0, compute capability: 5.2
TensorFlow:2.7.0, capabilities: [
	MKL,
	CUDA,
]
TensorFlow Library: C:\Users\java\.djl.ai\tensorflow\2.7.0-20220114-cu113-win-x86_64\jnitensorflow.dll
engine device = gpu(0)
dev=gpu(0)
model=Model (
	Name: mlp
	Data Type: null
)
Exception in thread "main" java.lang.UnsupportedOperationException: Not supported for TensorFlow Engine
	at ai.djl.tensorflow.engine.TfModel.setBlock(TfModel.java:158)
	at ai.djl.test.Test.main(Test.java:39)

`
public class Test {	
	public static void main(String[] args) {			
		Engine engine = new TfEngineProvider().getEngine(); 
		System.out.println(engine);
		
		Device[] devices = engine.getDevices();
		for (Device d: devices)
		{
			System.out.println("engine device = "+d); 			
		}		
		Device gpu = Device.gpu();
		System.out.println("dev="+gpu);		
		
		Model model = Model.newInstance("mlp",gpu,"TensorFlow");
		System.out.println("model="+model);
		model.setBlock(new SequentialBlock()); 
	}
}
`

**Comment by iamuspace** [comment]
TF cpu vesion on windows 10:  same error message 
2023-02-19 18:27:10.527183: I external/org_tensorflow/tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
TensorFlow:2.7.4, capabilities: [
	MKL,
]
TensorFlow Library: C:\Users\java\.djl.ai\tensorflow\2.7.4-20221103-cpu-win-x86_64\jnitensorflow.dll
engine device = cpu()
dev=cpu()
model=Model (
	Name: mlp
	Data Type: null
)
Exception in thread "main" java.lang.UnsupportedOperationException: Not supported for TensorFlow Engine
	at ai.djl.tensorflow.engine.TfModel.setBlock(TfModel.java:158)

**Comment by iamuspace** [comment]
pom.xml 


    <dependencies>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>bom</artifactId>
            <version>0.18.0</version>
            <type>pom</type>
        </dependency>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>djl-zero</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.java</groupId>
            <artifactId>java</artifactId>
            <version>0.18.0</version>
            <type>module</type>
        </dependency>
        <dependency>
            <groupId>com.google.protobuf</groupId>
            <artifactId>protobuf-java</artifactId>
            <version>3.19.2</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.android</groupId>
            <artifactId>pytorch-native</artifactId>
            <version>0.18.0</version>
            <type>module</type>
        </dependency>
        <dependency>
            <groupId>ai.djl.tensorflow</groupId>
            <artifactId>tensorflow-engine</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.tensorflow</groupId>
            <artifactId>tensorflow-api</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.tensorflow</groupId>
            <artifactId>tensorflow-native-cu113</artifactId>
            <version>2.7.0</version>
            <classifier>win-x86_64</classifier>
        </dependency>
    </dependencies>

**Comment by frankfliu** [comment, marked as answer]
@iamuspace 

In DJL, we only support inference for TensorFlow. You have to use PyTorch or MXNet if you want to train a model in DJL.

**Reply by iamuspace** [reply]
Thank you so much for your response. 
Your reply is very good "only support inference for TF". 

My env : 
Windows 10, AMD cpu 64bit. Nvidia Titan X GPU, Zulu 64bit JVM 
python tensorflow cuda is running OK. 

my DJL maven Java test result : 
pytorch : cpu (x), gpu (x) 
mxnet : cpu (o), gpu (x) 

I want to run cuda on my Windows 10 PC.
